Use keyword as datetime-diff units

metamben opened the pull request:
Fixes #29897.

<!-- Reviewable:start -->
- - -
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/metabase/metabase/30439)
<!-- Reviewable:end -->

metamben committed:
Use keyword as datetime-diff units

Fixes #29897.

bshepherdson reviewed (approved):
> Reviewed 3 of 3 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/metabase/metabase/30439)* status: :shipit: complete! all files reviewed, all discussions resolved (waiting on @metamben)



<!-- Sent from Reviewable.io -->

metamben committed:
Fix MLv2 round trip test

Fields with :base-type :type/Text are not considered temporal expressions
therefore :datetime-diff with such arguments is not a valid expression.
Such expressions removed by the clean step of the conversion to pMBQL.

metamben committed:
Merge remote-tracking branch 'origin/master' into 29897-fix-datetime-diff-schema

metamben committed:
Merge remote-tracking branch 'origin/master' into 29897-fix-datetime-diff-schema

metamben reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/metabase/metabase/30439)* status: 1 of 4 files reviewed, 1 unresolved discussion (waiting on @bshepherdson)

___
*[`test/metabase/query_processor_test/date_time_zone_functions_test.clj` line 1052 at r2](https://reviewable.io/reviews/metabase/metabase/30439#-NUMvSjq2oXt-4jFIaDq:-NUMvSjr-RVpMWsPdz59:b3lj2mi) ([raw file](https://github.com/metabase/metabase/blob/d9d4cc6d1d3bed69d9abe15eeef793bc17807a40/test/metabase/query_processor_test/date_time_zone_functions_test.clj#L1052)):*
> ```Clojure
>                                                                  [(name unit) [:datetime-diff
>                                                                                [:field "a_dt_tz" {:base-type :type/DateTime}]
>                                                                                [:field "b_dt_tz" {:base-type :type/DateTime}]
> ```

@camsaul Is this fix good enough, or should `:datetime-diff` work with fields of `:type/Text`? Or should we require a temporal `:effective-type` if `:base-type` is not temporal? Or something else?


<!-- Sent from Reviewable.io -->
